### ISSUE TITLE
How to modify the Faster R-CNN example to make it run on a dataset different from COCO

### ISSUE THREAD

**Issue by AndreaPi** [tensorpack/tensorpack#1189]
I want to apply the Mask R-CNN model to a model other than COCO, in order to perform **instance segmentation**:

http://dmery.ing.puc.cl/index.php/material/gdxray/

**If** I manage to make this work, I'll be glad to write a PR to add this as an example: the only problem is that I'm using Docker, which may complicate the process, but we can discuss this later if I succeed.

Now, the toy examples are explained very well, but the [Faster R-CNN](https://github.com/tensorpack/tensorpack/tree/master/examples/FasterRCNN) is, IMO, harder to follow (see also https://github.com/tensorpack/tensorpack/issues/1163). I have a few questions:

- how should I modify my dataset structure, so that I can reuse the Faster R-CNN example code with minimal modifications? Would something like this work?

```
GDXRAY/DIR/
  annotations/
    instances_train.json
    instances_val.json
  train/
    GDXRAY_train_*.jpg
  val/
    GDXRAY_val_*.jpg
```
 Where should I put the test data, in a folder `test/` at the same level as `train/` and `val/`?

- which is exactly the COCO format for the semantic segmentation task? The official page is http://cocodataset.org/#format-data . I guess it's `Object Detection`?

- what else should I modify, in order to make the Faster R-CNN example code run on a new dataset? For example, are there any "features" of the COCO datasets which are hard-coded in the Faster R-CNN example code, which I should be aware of?

Finally, a question which is not really Tensorpack-related: I'm asking anyway, just in case someone went through a similar hurdle and is willing to share some tips, but if it's not appropriate here, let me know and I'll remove this section. Are there tools which could speedup the process of writing annotations for my dataset? I'd rather not have to write each JSON file by hand. So far, I could find these resources:

https://github.com/jsbroks/imantics
http://www.immersivelimit.com/tutorials/create-coco-annotations-from-scratch/
https://patrickwasp.com/create-your-own-coco-style-dataset/
https://stackoverflow.com/questions/45545187/create-ms-coco-style-dataset
https://github.com/hanskrupakar/COCO-Style-Dataset-Generator-GUI

**Comment by ppwwyyxx** [comment]
> how should I modify my dataset structure, so that I can reuse the Faster R-CNN example code with minimal modifications? Would something like this work?

Yes. For minimal modifications to code, use COCO format.

> Where should I put the test data, in a folder test/ at the same level as train/ and val/?

There is no "testing" happen in the codebase (you would manually provide path to input images for testing).
If you want to add a new dataset for evaluation, it can be in the same format as "val" (i.e., a `instances_XXX.json` and `XXX/*.jpg`

> which is exactly the COCO format for the semantic segmentation task? The official page is http://cocodataset.org/#format-data . I guess it's Object Detection?

Yes. You can also download one of the json there (pick a small one) and take a look.

> what else should I modify, in order to make the Faster R-CNN example code run on a new dataset? For example, are there any "features" of the COCO datasets which are hard-coded in the Faster R-CNN example code, which I should be aware of?

All you need to change are:
1. `cfg.DATA.TRAIN/VAL/NUM_CATEGORY`
2. `COCO_id_to_category_id` and `class_names` in `COCODetection`.

**Comment by ppwwyyxx** [comment]
One thing I forgot: register the name of your datasets here: https://github.com/tensorpack/tensorpack/blob/f0626dcb826e9027295bada03033bbfaef5008b3/examples/FasterRCNN/dataset/coco.py#L205-L212
This was recently added to map names into datasets.

Also, I realized that modifying `COCO_id_to_category_id` is confusing to people unfamiliar with COCO, so after https://github.com/tensorpack/tensorpack/commit/2902bfbe018a753543bc84ac2cf44f015c5aa1b0 the dict can be made an empty dict for your datasets.

**Comment by AndreaPi** [comment]
> There is no "testing" happen in the codebase (you would manually provide path to input images for testing).
If you want to add a new dataset for evaluation, it can be in the same format as "val" (i.e., a `instances_XXX.json` and `XXX/*.jpg`

Ok, you say I can manually provide a path to input images for testing. Since I want to compute the test loss on a test dataset with multiple images, is it possible to provide an entire folder for testing? Or do I have to test on a single image at a time?

**Comment by ppwwyyxx** [comment]
> Since I want to compute the test loss on a test dataset with multiple images

There is no definition of a "test loss" in R-CNN unless you define it somehow first. You can compute a training loss on test dataset (if you have preprocessed its annotations in the same way as training).

If you just want to compute predictions on test images, it's in the README https://github.com/tensorpack/tensorpack/tree/master/examples/FasterRCNN#inference and support multiple images. You can change the names at https://github.com/tensorpack/tensorpack/blob/8879637399b665bfd11aa5ded663619a1a19bdb9/examples/FasterRCNN/predict.py#L140 to let it compute any other tensors defined in the graph. (But again, there is no "loss" defined in the testing graph).

**Comment by AndreaPi** [comment]
> There is no definition of a "test loss" in R-CNN unless you define it somehow first. You can compute a training loss on test dataset (if you have preprocessed its annotations in the same way as training).

This is very interesting. I'm indeed annotating the test dataset the same way I'm annotating train and val. But I'm a bit surprised you cannot compute a loss on the test set - I understood [a few papers reported the mAP](https://medium.com/@jonathan_hui/object-detection-speed-and-accuracy-comparison-faster-r-cnn-r-fcn-ssd-and-yolo-5425656ae359) [on COCO Testing images](http://images.cocodataset.org/zips/test2014.zip). I thought I could compute mAP on a test dataset with

```
./predict.py --evaluate output.json --load /path/to/Trained-Model-Checkpoint \
    --config SAME-AS-TRAINING
```
but maybe this evaluates on the training set?

 Either I misunderstand something, or it's possible to compute the mAP but not the loss. This is curious. Let me know if you'd be available to discuss the topic more on a channel of your choice (Slack, email, Gitter, etc.) and I'll send you my contacts, otherwise don't worry and thanks a lot for your support.

**Comment by ppwwyyxx** [comment]
The script does compute the mAP on `cfg.DATA.VAL` but not the loss.

As far as I know there is no definition on what is "test loss" for R-CNN. To some people this may mean exactly the "training loss" computed on validation dataset. But it may mean something else to others.
We can discuss at tensorpack gitter (https://gitter.im/tensorpack/users) if you like.
